bindings/resolver.go

```diff
@@ -0,0 +1,253 @@
+package bindings
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// SecretReader represents the read-only content of a secret value.
+type SecretReader struct {
+	data *bytes.Buffer
+}
+
+// MakeSecretReader returns a SecretReader from the given secret data.
+func MakeSecretReader(data []byte) SecretReader {
+	return SecretReader{
+		data: bytes.NewBuffer(data),
+	}
+}
+
+// Bytes returns the raw content of the secret.
+func (r SecretReader) Bytes() []byte {
+	return r.data.Bytes()
+}
+
+// String returns the content of the secret as a string.
+func (r SecretReader) String() string {
+	return r.data.String()
+}
+
+// Read reads the next len(p) bytes from the secret or until the secret is drained. The return value n is the number of
+// bytes read. If the secret has no data to return, err is io.EOF (unless len(p) is zero); otherwise it is nil.
+func (r SecretReader) Read(p []byte) (int, error) {
+	return r.data.Read(p)
+}
+
+// Binding represents metadata related to an external service.
+type Binding struct {
+
+	// Name is the name of the binding, given by the binding directory name.
+	Name string
+
+	// Path is the path to the binding directory.
+	Path string
+
+	// Type is the type of the binding, given by the content of the 'type' file within the binding directory.
+	Type string
+
+	// Provider is the provider of the binding, given by the content of the 'provider' file within the binding
+	// directory.
+	Provider string
+
+	// Secret is the primary content of the binding. Keys are given by each file name within the binding directory
+	// (other than 'type' or 'provider'), and corresponding values are given by the content of each file.
+	Secret map[string]SecretReader
+}
+
+// Resolver resolves service bindings according to the kubernetes binding spec:
+// https://github.com/k8s-service-bindings/spec#workload-projection.
+type Resolver struct {
+	bindingRoot string
+	bindings    []Binding
+}
+
+// NewResolver returns a new service binding resolver. The location of the bindings is given by one of the following, in
+// the following order of precedence:
+//
+//  1. SERVICE_BINDING_ROOT environment variable
+//  2. CNB_BINDINGS environment variable, if above is not set
+//  3. `<platformDir>/bindings`, if both above are not set
+func NewResolver(platformDir string) *Resolver {
+	root := os.Getenv("SERVICE_BINDING_ROOT")
+	if root == "" {
+		root = os.Getenv("CNB_BINDINGS")
+	}
+
+	if root == "" {
+		root = filepath.Join(platformDir, "bindings")
+	}
+
+	return &Resolver{
+		bindingRoot: root,
+	}
+}
+
+// Resolve returns all bindings matching the given type and optional provider (case-insensitive). To match on type only,
+// provider may be an empty string. Returns an error if there are problems loading bindings from the file system.
+func (r *Resolver) Resolve(typ string, provider string) ([]Binding, error) {
```

Comment:
I know that it's the convention for interface naming (e.g. `Stringer.String()`), but I understand that it may not be as desirable for structs. I just don't know why and have no personal reason for avoiding it (other than sometimes name collisions with interfaces). It also seems hard to avoid when trying to come up with descriptive names.

I'd personally prefer to leave as is, if it's just a nit.

But if you feel more strongly, I'm not opposed to renaming. I think that the names above are perhaps ambiguous as to the plurality (mostly on `Get` as it doesn't necessarily imply one), so if I were to change this, I might suggest something like `Get` and `GetOne`, or `Filter` and `FilterOne`, or something like that.